src/rust/include/rust/orchard/keys.h

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) 2021 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php .
+
+#ifndef ZCASH_RUST_INCLUDE_RUST_ORCHARD_KEYS_H
+#define ZCASH_RUST_INCLUDE_RUST_ORCHARD_KEYS_H
+
+#include "rust/streams.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * A pointer to an Orchard shielded payment address, as defined in
+ * https://zips.z.cash/protocol/nu5.pdf#orchardpaymentaddrencoding
+ */
+struct OrchardRawAddressPtr;
+typedef struct OrchardRawAddressPtr OrchardRawAddressPtr;
+
+/**
+ * Clones the given Orchard payment address and returns
+ * a pointer to the newly created value. Both the original
+ * one's memory and the newly allocated one need to be freed
+ * independently.
+ */
+OrchardRawAddressPtr* orchard_address_clone(
+        const OrchardRawAddressPtr* ptr);
+
+/**
+ * Frees the memory allocated for the given Orchard payment address
+ */
+void orchard_address_free(OrchardRawAddressPtr* ptr);
+
+//
+// Incoming Viewing Keys
+//
+
+struct OrchardIncomingViewingKeyPtr;
+typedef struct OrchardIncomingViewingKeyPtr OrchardIncomingViewingKeyPtr;
+
+/**
+ * Clones the given Orchard incoming viewing key and returns
+ * a pointer to the newly created value. Both the original
+ * one's memory and the newly allocated one need to be freed
+ * independently.
+ */
+OrchardIncomingViewingKeyPtr* orchard_incoming_viewing_key_clone(
+        const OrchardIncomingViewingKeyPtr* ptr);
+
+/**
+ * Frees the memory allocated for the given Orchard incoming viewing key
+ */
+void orchard_incoming_viewing_key_free(OrchardIncomingViewingKeyPtr* ptr);
+
+/**
+ * Returns the address at the specified diversifier index.
+ */
+OrchardRawAddressPtr* orchard_incoming_viewing_key_to_address(
+    const OrchardIncomingViewingKeyPtr* incoming_viewing_key,
+    const unsigned char* j);
+
+/**
+ * Parses an Orchard incoming viewing key from the given stream.
+ *
+ * - If the key does not parse correctly, the returned pointer will be null.
+ */
+OrchardIncomingViewingKeyPtr* orchard_incoming_viewing_key_parse(
+    void* stream,
+    read_callback_t read_cb);
+
+/**
+ * Serializes an Orchard incoming viewing key to the given stream
+ *
+ * If `incoming_viewing_key == nullptr`, this serializes `nActionsOrchard = 0`.
```

Comment:
`incoming_viewing_key` isn't allowed to be null, and this comment is left over from copy-pasta:
```suggestion
 * Serializes an Orchard incoming viewing key to the given stream.
```